code/datums/components/pitfall.dm

```diff
@@ -0,0 +1,199 @@
+ * 	// this is the code for falling from abyss into ice caves
+ *	// could maybe use an animation, or better text. perhaps a slide whistle ogg?
+ **/
+
+ABSTRACT_TYPE(/datum/component/pitfall)
+/// A component for turfs which make movable atoms "fall down a pit"
+/datum/component/pitfall
+	/// a list of targets for the fall to pick from
+	var/list/TargetList = list()
+	/// the maximum amount of brute damage applied. This is used in random_brute_damage()
+	var/BruteDamageMax = 0
+	/// How long it takes for a thing to fall into the pit. 0 is instant, but usually you'd have a couple deciseconds where something can be flung across. Should use time defines.
+	var/FallTime = 0.3 SECONDS
+
+	// --------------- landmark targeting
+	/// The landmark that the fall sends you to. Should be a landmark define.
+	var/TargetLandmark = ""
+
+	// --------------- area targeting
+	/// The area path that the target falls into. For area targeting
+	var/TargetArea = null
+
+	// --------------- coordinate targeting
+	/// The z level that the target falls into if not via area or landmark.
+	var/TargetZ = 5
+	/// If truthy, try to find a spot around the target to land on in range(x).
+	var/LandingRange = 8
+
+
+/datum/component/pitfall/Initialize(BruteDamageMax = 50, FallTime = 0.3 SECONDS)
+	. = ..()
+	if (!istype(src.parent, /turf))
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(src.parent, COMSIG_ATOM_ENTERED, PROC_REF(start_fall))
+	RegisterSignal(src.parent, COMSIG_ATTACKBY, PROC_REF(update_targets))
+	src.BruteDamageMax	= BruteDamageMax
+	src.FallTime		= FallTime
+	src.update_targets()
+
+
+/// returns the .parent but typecasted as a turf
+/datum/component/pitfall/proc/typecasted_parent()
+	RETURN_TYPE(/turf)
+	. = src.parent
+
+/// updates targets for area/coordinate targeting. is overridden added to in child types.
+/datum/component/pitfall/proc/update_targets()
+	return
+
+/// called when movable atom AM enters a pitfall turf.
+/datum/component/pitfall/proc/start_fall(var/signalsender, var/atom/movable/AM)
+	if (!istype(AM, /atom/movable) || istype(AM, /datum/projectile/))
+		return
+	if (HAS_FLAG(AM.event_handler_flags, IMMUNE_TRENCH_WARP))
+		return
+	if (AM.anchored || locate(/obj/lattice) in src.parent)
+		return
+	if (ismob(AM))
+		var/mob/M = AM
+		if (M.client?.flying || isobserver(AM) || isintangible(AM))
+			return
+
+	return_if_overlay_or_effect(AM)
+
+	if (src.FallTime)
+		SPAWN(src.FallTime)
+			AM.loc.GetComponent(/datum/component/pitfall)?.try_fall(signalsender, AM)
+	else
+		src.try_fall(signalsender, AM)
+
+/// called when it's time for movable atom AM to actually fall into the pit
+/datum/component/pitfall/proc/try_fall(var/signalsender, var/atom/movable/AM)
+	if (istype(AM, /obj/machinery/vehicle))
+		var/obj/machinery/vehicle/V = AM
+		var/turf/target_turf = V.go_home()
+		if (V.going_home && target_turf)
+			V.going_home = 0
+			AM.set_loc(target_turf)
+			return
+
+	if (!src.TargetList || !length(src.TargetList))
+		src.update_targets()
```

Comment:
Yeah basically here I'd just call some `var/turf/target = src.get_target_turf(AM)` proc and then fall to that target. Those subtypes which need a TargetList can update it if they want. Having a separate proc to get the target turf allows for more flexibility (for example it avoids the current awkward workaround for the landmark stuff).

Reply:
i've moved it to the z level fall code section, but i've had to make it a little messy.